lib/filterx/filterx-variable.h

```diff
@@ -117,6 +117,7 @@ filterx_variable_set_value(FilterXVariable *v, FilterXObject **new_value, gboole
   filterx_object_unref(v->value);
   if (new_value)
     {
+      *new_value = filterx_ref_new(*new_value);
```

Comment:
`filterx_ref_new()` avoids nesting refs in refs or wrapping non-mutable objects, so this won't have too big of an overhead, but my original idea was to create refs only when necessary, when we see
- an assignment
- setsubscript, setattr
- when dealing with dict/list entries internally

I implemented this on the expression level, because something was not right when I tried it on the variable layer.

Generally, we want all mutable objects to be handler through refs, as you mentioned. But if we add this here, can't we remove a few from the expression side?

Reply:
I don't think we can, but we can still discuss this. Right now we can wrap objects into refs at three places:

1) as a result of an expression, it can already be boxed (for instance create_container for generators)
2) right before an assignment/setattr/set_subscript
3) right before storing it into a mutable container, such as a dict/list

This makes sure that eventually everything is boxed into a ref, even if steps prior did not do it. I think we might be able to remove some of this if we get rid off generators.